Add tutorial about running Pods with sidecar containers

SergeyKanzhelev opened the pull request:
/sig node

more questions about sidecars warranted me to write this down. 

It was discussed at Sidecar WG meeting: https://docs.google.com/document/d/1E1guvFJ5KBQIGcjCrQqFywU9_cBQHRtHvjuqcVbCXvU/edit#bookmark=id.u4zmu7j9vn31


This is the first step in this tutorial. It may require more work in the following areas:

1. I am out of my depth on what would be the best practices writing tools and mutating web hooks that will preserve unknown fields: https://github.com/kubernetes/website/pull/46825#discussion_r1710243472 Extended documentation or even a link to other page that may explain this would go a long way. (https://github.com/kubernetes/website/issues/47465)
2. Explain how to implement sidecar containers without the built-in feature. I am not sure how deep we want to go there to cover edge cases. Perhaps just cover the very basic scenario.
3. Fully working example of a sidecar container that handles termination properly would go a long way explaining concepts.

tengqm commented on content/en/docs/concepts/workloads/pods/sidecar-containers.md:
```diff
@@ -141,6 +141,80 @@ limit.
+
+This section is relevant for people adopting sidecar containers for their workloads.
+
+The sidecar containers feature is in beta state starting Kubernetes version 1.29 and is enabled by default.
```
Better use the shortcode for feature gates.
This sentence becomes invalid when the gate is removed.
By the way, "sidecar containers" is not the name of the gate.

tengqm commented on content/en/docs/concepts/workloads/pods/sidecar-containers.md:
```diff
@@ -141,6 +141,80 @@ limit.
 On Linux, resource allocations for Pod level control groups (cgroups) are based on the effective Pod
 request and limit, the same as the scheduler.
 
+### Adopting sidecar containers
```
This section might be more suitable for a reference page rather than a section of the concepts page.

SergeyKanzhelev replied:
I would appreciate a pointer on which page is better. I didn't think it deserve the whole task page under https://kubernetes.io/docs/tasks/. I am not sure where under reference would be a good place.

tengqm commented on content/en/docs/concepts/workloads/pods/sidecar-containers.md:
```diff
@@ -141,6 +141,80 @@ limit.
+**1. Ensure the feature gate is enabled.**
+
+As a very first step, make sure that both - API server and Nodes were upgraded to the version 1.29+.
+The feature will break on clusters where Nodes are running earlier versions where it is not enabled.
```
Is it okay to run a 1.28 version Node with the feature (still Alpha) enabled?

SergeyKanzhelev replied:
added an alert below

tengqm commented on content/en/docs/concepts/workloads/pods/sidecar-containers.md:
```diff
@@ -141,6 +141,80 @@ limit.
+
+The feature gate enablement should also be validated for both - API server AND Nodes.
+
+Once of the ways to check the feature gate enablement is running a command like this:
```
Once -> One

SergeyKanzhelev replied:
```suggestion
One of the ways to check the feature gate enablement is running a command like this:
```

tengqm commented on content/en/docs/concepts/workloads/pods/sidecar-containers.md:
```diff
@@ -141,6 +141,80 @@ limit.
+- For API Server 
+  `kubectl get --raw /metrics | grep kubernetes_feature_enabled | grep SidecarContainers`
+- For the individual node: 
+  `kubectl get --raw /api/v1/nodes/gke-cluster-device-p-almost-real-tpu--122d9b42-tnb5/proxy/metrics | grep kubernetes_feature_enabled | grep SidecarContainers`
```
```suggestion
  `kubectl get --raw /api/v1/nodes/<node-name>/proxy/metrics | grep kubernetes_feature_enabled | grep SidecarContainers`
```

tengqm commented on content/en/docs/concepts/workloads/pods/sidecar-containers.md:
```diff
@@ -141,6 +141,80 @@ limit.
+
+**2. Check for 3rd party tooling and mutating webhooks**
+
+If trying the feature you experience issues, it may be an indication that one of the
```
```suggestion
If you experience issues when validating the feature, it may be an indication that one of the
```

tengqm commented on content/en/docs/concepts/workloads/pods/sidecar-containers.md:
```diff
@@ -141,6 +141,80 @@ limit.
+If trying the feature you experience issues, it may be an indication that one of the
+3rd party tools or mutating webhooks are broken.
+
+The Sidecar feature adds a new field that is define in Kubernetes 1.28 API.
```
```suggestion
The SidecarContainers feature adds a new field that is define in Kubernetes 1.28 API.
```

tengqm commented on content/en/docs/concepts/workloads/pods/sidecar-containers.md:
```diff
@@ -141,6 +141,80 @@ limit.
+3rd party tools or mutating webhooks are broken.
+
+The Sidecar feature adds a new field that is define in Kubernetes 1.28 API.
+Some tools and mutating webhooks might have been build with the earlier version of Kubernetes API.
```
```suggestion
Some tools or mutating webhooks might have been built with an earlier version of Kubernetes API.
```

tengqm commented on content/en/docs/concepts/workloads/pods/sidecar-containers.md:
```diff
@@ -141,6 +141,80 @@ limit.
+Some tools and mutating webhooks might have been build with the earlier version of Kubernetes API.
+If tools pass the unknown fields as-is or mutating webhooks are using the recommended `Patch` method to mutate Pods, it will not be a problem.
+
+However there are tooling that will strip out unknown fields and must be recompiled with the fresh version of Kubernetes API.
```
```suggestion
However there are tools that will strip out unknown fields and must be recompiled with the v1.28+ version of Kubernetes API.
```

tengqm commented on content/en/docs/concepts/workloads/pods/sidecar-containers.md:
```diff
@@ -141,6 +141,80 @@ limit.
+
+However there are tooling that will strip out unknown fields and must be recompiled with the fresh version of Kubernetes API.
+
+The easy way to validate if this is a root cause of a problem is to use `kiubectl describe pod` command with your Pod.
```
```suggestion
The way to check this is to use the `kubectl describe pod` command with your Pod.
```

tengqm commented on content/en/docs/concepts/workloads/pods/sidecar-containers.md:
```diff
@@ -141,6 +141,80 @@ limit.
+However there are tooling that will strip out unknown fields and must be recompiled with the fresh version of Kubernetes API.
+
+The easy way to validate if this is a root cause of a problem is to use `kiubectl describe pod` command with your Pod.
+If any tooling stripped out the new field (`restartPolicy:Always`), you will not find see it in the command output.
```
```suggestion
If any tools stripped out the new field (`restartPolicy:Always`), you will not find see it in the command output.
```

sftim replied:
Even better: avoid “new” here. We like [timeless documentation](https://developers.google.com/style/timeless-documentation).

tengqm commented on content/en/docs/concepts/workloads/pods/sidecar-containers.md:
```diff
@@ -141,6 +141,80 @@ limit.
+The easy way to validate if this is a root cause of a problem is to use `kiubectl describe pod` command with your Pod.
+If any tooling stripped out the new field (`restartPolicy:Always`), you will not find see it in the command output.
+
+If you hit an issue like this, please also advice authors of tooling or web hooks to switch to `Patch` method of modifying objects.
```
```suggestion
If you hit an issue like this, please advise the author of the tools or the webhooks to switch to `Patch` method of modifying objects.
```

tengqm commented on content/en/docs/concepts/workloads/pods/sidecar-containers.md:
```diff
@@ -141,6 +141,80 @@ limit.
+If you are using software that injects sidecars automatically,
+there are a few possible strategies you may follow to
+ensure that native sidecar container can be used.
+All of the strategies are generally a ways you may choose to decide whether
```
```suggestion
All of the strategies are generally options you may choose to decide whether
```

sftim reviewed (commented):
It's great that we want to help people learn about running sidecars.

How about adding a tutorial page, eg https://k8s.io/docs/tutorials/pod-sidecar-containers/?

SergeyKanzhelev commented:
> It's great that we want to help people learn about running sidecars.
> 
> How about adding a tutorial page, eg https://k8s.io/docs/tutorials/pod-sidecar-containers/?

Updated.

sftim commented on content/en/docs/tutorials/configuration/pod-sidecar-containers.md:
```diff
@@ -0,0 +1,114 @@
+
+{{< include "task-tutorial-prereqs.md" >}} {{< version-check >}}
+
+* This tutorial is relevant for clusters version 1.29+
```
The `version-check` shortcode can cover this for you.

sftim replied:
```suggestion
```

sftim commented on content/en/docs/tutorials/configuration/pod-sidecar-containers.md:
```diff
@@ -0,0 +1,114 @@
+
+This section is relevant for people adopting [sidecar containers](/docs/concepts/workloads/pods/sidecar-containers/) for their workloads.
+
+{{< feature-state for_k8s_version="v1.29" state="beta" >}}
```
```suggestion
{{< feature-state feature_gate_name=`SidecarContainers >}}
```

sftim commented on content/en/docs/tutorials/_index.md:
```diff
@@ -27,6 +27,7 @@ Before walking through each tutorial, you may want to bookmark the
 
 * [Example: Configuring a Java Microservice](/docs/tutorials/configuration/configure-java-microservice/)
 * [Configuring Redis Using a ConfigMap](/docs/tutorials/configuration/configure-redis-using-configmap/)
+* [Adopting Sidecar Containers](/docs/tutorials/configuration/pod-sidecar-containers/)
```
This is not a configuration tutorial. Try “Stateless applications”?

sftim commented on content/en/docs/concepts/workloads/pods/sidecar-containers.md:
```diff
@@ -147,3 +147,4 @@ request and limit, the same as the scheduler.
 * Read about [creating a Pod that has an init container](/docs/tasks/configure-pod-container/configure-pod-initialization/#create-a-pod-that-has-an-init-container).
 * Learn about the [types of probes](/docs/concepts/workloads/pods/pod-lifecycle/#types-of-probe): liveness, readiness, startup probe.
 * Learn about [pod overhead](/docs/concepts/scheduling-eviction/pod-overhead/).
+* Learn how to [Adopt Sidecar Containers](/docs/tutorials/configuration/pod-sidecar-containers/)
```
I'd list this as the first item in the list, not the last/

SergeyKanzhelev replied:
makes sense. moved

sftim commented on content/en/docs/tutorials/configuration/pod-sidecar-containers.md:
Overall feedback:
- Explain the meaning of sidecar containers; encourage people to read the concept guide first, but don't assume that they actually did.
  - Avoid assuming that readers know what Pods are. For example, they want to run 17 replicas of a container on a cluster with 10 nodes, where it so happens one of those nodes doesn't have any replica running. They might calculate 10 - 1 = 9 and think that they'd need 9 Pods. A very brief outline of what Pods are would really help, plus a link to the concept page.
- We should teach that you can run two app containers side by side but that using a sidecar that Kubernetes recognizes brings its own benefits
- We ought to explain init containers (that run to completion) first, then show how to keep an init container running throughout the lifetime of an app container. And we should explain a little about why we're teaching it this way.
- If using `{{< version-check >}}`, set `min-kubernetes-server-version` in front matter
- “sidecar containers” not “Sidecar Containers”
- don't number headings

SergeyKanzhelev replied:
I updated text. Would you say it is enough depth or I should go deeper in explaining concepts in this tutorial?

sftim replied:
See https://github.com/kubernetes/website/pull/46825#pullrequestreview-2134319796

SergeyKanzhelev commented:
Any advice what does this may mean:

> Error: error building site: assemble: "/opt/build/repo/content/en/docs/tutorials/configuration/pod-sidecar-containers.md:1:1": got positional parameter 'or'. Cannot mix named and positional parameters

sftim commented on content/en/docs/tutorials/configuration/pod-sidecar-containers.md:
```diff
@@ -0,0 +1,148 @@
+
+This section is relevant for people adopting [sidecar containers](/docs/concepts/workloads/pods/sidecar-containers/) for their workloads.
+
+{{< feature-state feature_gate_name=`SidecarContainers >}}
```
```suggestion
{{< feature-state feature_gate_name="SidecarContainers" >}}
```
should fix it.

sftim reviewed (commented):
> Any advice what does this may mean:
> 
> > ```
> > Error: error building site: assemble: "/opt/build/repo/content/en/docs/tutorials/configuration/pod-sidecar-containers.md:1:1": got positional parameter 'or'. Cannot mix named and positional parameters
> > ```

See inline feedback (the `feature-state` suggestion)

sftim commented on content/en/docs/tutorials/configuration/pod-sidecar-containers.md:
```diff
@@ -0,0 +1,148 @@
+
+The `SidecarContainers` feature adds a new field that is define in Kubernetes `1.28` API.
+Some tools or mutating webhooks might have been built with an earlier version of Kubernetes API.
+If tools pass the unknown fields as-is or mutating webhooks are using the recommended `Patch` method to mutate Pods, it will not be a problem.
```
Maybe you meant:
```markdown
-If tools pass the unknown fields as-is or mutating webhooks are using the recommended `Patch` method to mutate Pods, it will not be a problem.
+If tools pass the unknown fields as-is or mutating webhooks are using the recommended `PATCH` method
+to mutate Pods, it will not be a problem.
```
HTTP verbs are typically all-uppercase. However, **mutating webhooks don't use PATCH** because they are an HTTP callout (and response) that modifies the request body during admission.

SergeyKanzhelev replied:
I rephrased

sftim commented on content/en/docs/tutorials/configuration/pod-sidecar-containers.md:
```diff
@@ -0,0 +1,148 @@
+Some tools or mutating webhooks might have been built with an earlier version of Kubernetes API.
+If tools pass the unknown fields as-is or mutating webhooks are using the recommended `Patch` method to mutate Pods, it will not be a problem.
+
+However there are tools that will strip out unknown fields and must be recompiled with the v1.28+ version of Kubernetes API.
```
```suggestion
However there are tools that will strip out unknown fields; if you have those, they must be recompiled with the v1.28+ version of Kubernetes API client code.
```

sftim commented on content/en/docs/tutorials/configuration/pod-sidecar-containers.md:
```diff
@@ -0,0 +1,148 @@
+
+However there are tools that will strip out unknown fields and must be recompiled with the v1.28+ version of Kubernetes API.
+
+The way to check this is to use the `kubectl describe pod` command with your Pod.
```
```suggestion
The way to check this is to use the `kubectl describe pod` command with your Pod that has passed through
mutating admission.
```

sftim commented on content/en/docs/tutorials/configuration/pod-sidecar-containers.md:
```diff
@@ -0,0 +1,148 @@
+The way to check this is to use the `kubectl describe pod` command with your Pod.
+If any tools stripped out the new field (`restartPolicy:Always`), you will not find see it in the command output.
+
+If you hit an issue like this, please advise the author of the tools or the webhooks to switch to `Patch` method of modifying objects.
```
I don't understand this advice. Webhooks do not use HTTP `PATCH`. Do you mean `"patchType": "JSONPatch"`?

The thing is, [`patchType`](https://kubernetes.io/docs/reference/config-api/apiserver-admission.v1/#admission-k8s-io-v1-PatchType) can only ever be JSON Patch; no other type is supported.

If we are writing advice for authors of API clients that do issue HTTP `PATCH` requests, are we sure that Server Side Apply or JSON Merge Patch would be unsuitable?

Please clarify the point you want to make in a review comment, and also reword accordingly?

sftim replied:
We can provide additional advice that is specific to people coding in Go; however, the initial / baseline advice should be agnostic about the implementation language for the webhook. We don't enforce or recommend an implementation language for admission-phase extensions.

SergeyKanzhelev replied:
this is very tricky to give a proper recommendation. We see that `jsonpatch.CreatePatch` sometimes decides to generate a patch that instead of injecting a full new container, will re-inject what was the sidecar container (without the restartPolicy) and then change some fields on what was the sidecar container instead.

Container:
```
apiVersion: v1
kind: Pod
metadata:
  name: asm-sidecar-test
  namespace: default
spec:
  initContainers:
  - name: busybox-native-sidecar
    image: busybox
    restartPolicy: Always
    command:
      - "/bin/sh"
      - "-c"
      - |
        sleep infinite
  containers:
  - name: busybox
    image: busybox
    command:
      - "/bin/sh"
      - "-c"
      - |
        sleep infinite
```

Patch:

```
  {
    "op": "add",
    "path": "/spec/initContainers/1",
    "value": {
      "name": "busybox-native-sidecar",
      "image": "busybox",
      "command": [
        "/bin/sh",
        "-c",
        "sleep infinite\n"
      ],
      "resources": {}
    }
  },
  {
    "op": "replace",
    "path": "/spec/initContainers/0/name",
    "value": "istio-validation"
  },
  {
    "op": "replace",
    "path": "/spec/initContainers/0/image",
    "value": "gcr.io/istio-testing/proxyv2:latest"
  },
```

So the recommendation is basically manually craft the fields mutation

SergeyKanzhelev replied:
I made the very generic statement in the doc, I am going out of my depth here. Left a comment in PR description that this is a great follow up to have. Perhaps even a separate document as it come up often in KEP's PRRs - how to handle new fields being removed.

sftim commented on content/en/docs/tutorials/configuration/pod-sidecar-containers.md:
```diff
@@ -0,0 +1,148 @@
+## {{% heading "whatsnext" %}}
+
+
+* Learn more about [SidecarContainers](/docs/concepts/workloads/pods/sidecar-containers/).
```
```suggestion
* Learn more about [sidecar containers](/docs/concepts/workloads/pods/sidecar-containers/).
```

sftim reviewed (changes requested):
Questions in my mind:
- After following all the steps in the tutorial, how many Pods are still running (did we explain how to clean up?)
- After following all the steps in the tutorial, how many Pods that have a sidecar container have started and become healthy?
- After following all the steps in the tutorial, how many other Pods started (eg to help make a point about failure handling).

If we're not helping the reader to run any Pods that have sidecars, the tutorial needs expanding until that changes.

sftim commented on content/en/docs/tutorials/configuration/pod-sidecar-containers.md:
```diff
@@ -0,0 +1,148 @@
+As a very first step, make sure that both - API server and Nodes were upgraded to the version `1.29+`.
+The feature will break on clusters where Nodes are running earlier versions where it is not enabled.
+
+The feature gate enablement should also be validated for both - API server AND Nodes.
```
(we can use Markdown for emphasis, eg bold text)

sftim commented:
/retitle Add tutorial about running Pods with sidecar containers

network-charles commented on content/en/docs/tutorials/configuration/pod-sidecar-containers.md:
```diff
@@ -0,0 +1,148 @@
+
+Sidecar containers are the secondary containers that run along with the main
+application container within the same {{< glossary_tooltip text="Pod" term_id="pod" >}}.
+These containers are used to enhance or to extend the functionality of the primary _app
+container_ by providing additional services, or functionality such as logging, monitoring,
+security, or data synchronization, without directly altering the primary application code.
```
Thanks for opening this PR. This suggestion achieves the following:
- It removes the redundant "to" between the two verbs "enhance" and "extend".
- It breaks the sentence into two parts to improve clarity. Shorter sentences are easier to read and understand.
- it changes the singular noun "functionality" to a plural "functionalities". Using a plural denotes that each additional service is distinct, but they contribute to the app's overall functionality.

```suggestion
These containers are used to enhance or extend the functionality of the primary _app
container_. They provide additional services, or functionalities such as logging, monitoring,
security, or data synchronization, without directly altering the primary application code.
```

network-charles commented on content/en/docs/tutorials/configuration/pod-sidecar-containers.md:
```diff
@@ -0,0 +1,148 @@
+There are many benefits for the built-in sidecar containers.
+
+1. Built-in sidecar containers can run before Init containers.
+2. It is guaranteed that built-in sidecar containers finish AFTER all the regular containers completion.
```
The statement is correct, but it can easily be misinterpreted. At first read, I assumed you meant that the sidecar container would finish successfully with a zero exit code. However, when I tried it by making it run continuously, it finished with a non-zero exit code.

To try this, define a manifest, set the regular container to sleep for say, `10s`, and set the pod’s restart policy to `NEVER`. Lastly, set the restart policy of the init-container to `ALWAYS` and make sure it keeps running by using a while loop that starts it after the sleep of `1s` elapses.
```yaml
apiVersion: v1
kind: Pod
metadata:
  name: my-pod
spec:
  initContainers:
  - name: sidecar
    image: alpine
    command: ["/bin/sh", "-c", "while true; do echo 'Sidecar is running'; sleep 1; done"] 
    restartPolicy: Always
  containers:
  - name: main
    image: alpine
    command: ["/bin/sh", "-c", "echo 'Main container is running'; sleep 10"]
  restartPolicy: Never
```

You will notice that the main container completes successfully and terminates. However, the sidecar container keeps running until it is forcefully terminated via the `SIGKILL` signal, hence, causing the pod to be at a pod phase of `Succeeded` or a status of `Completed` when you do a `kubectl get pod`.

```suggestion
2. It is guaranteed that built-in sidecar containers are terminated once all the regular containers are completed and terminated.
```

sftim replied:
Sidecars do get a shutdown signal AIUI, but `["/bin/sh", "-c", "while true; do echo 'Sidecar is running'; sleep 1; done"]` might not have a signal handler that will pass the shutdown signal to `sleep` in a useful way.

ICBW here.

network-charles replied:
I agree. If you use an nginx image for the sidecar and main container, but add a short sleep command only for the main container, both containers shut down successfully with an exit code of 0. 

**pod.yaml**
```yaml
apiVersion: v1
kind: Pod
metadata:
  name: my-pod
spec:
  initContainers:
  - name: sidecar
    image: nginx:alpine
    restartPolicy: Always
  containers:
  - name: main
    image: alpine
    command: ["/bin/sh", "-c", "echo 'Main container is running'; sleep 10"]
  restartPolicy: Never
```

```yaml
kubectl apply -f pod.yaml

NAME     READY   STATUS    RESTARTS   AGE
my-pod   2/2     Running   0          11s
```

After about 14 seconds
```yaml
kubectl get pod
NAME     READY   STATUS      RESTARTS   AGE
my-pod   0/2     Completed   0          15s
```

```yaml
kubectl logs my-pod sidecar 

2024/07/01 18:06:15 [notice] 1#1: start worker processes
2024/07/01 18:06:15 [notice] 1#1: start worker process 29
2024/07/01 18:06:26 [notice] 1#1: signal 3 (SIGQUIT) received, shutting down
2024/07/01 18:06:26 [notice] 29#29: gracefully shutting down
2024/07/01 18:06:26 [notice] 29#29: exiting
2024/07/01 18:06:26 [notice] 29#29: exit
2024/07/01 18:06:26 [notice] 1#1: signal 17 (SIGCHLD) received from 29
2024/07/01 18:06:26 [notice] 1#1: worker process 29 exited with code 0
2024/07/01 18:06:26 [notice] 1#1: exit
```

```yaml
kubectl describe pod my-pod

main:
    Container ID:  containerd://d67d484d263b88f954b110898b9d0a81b6ff8657143d7a3043f07c4a5a518d3b
    Image:         alpine
....
    Command:
      /bin/sh
      -c
      echo 'Main container is running'; sleep 10
    State:          Terminated
      Reason:       Completed
      Exit Code:    0
```

You can see that they both handle the shutdown signal gracefully.

I am only saying that the statement in the commit saying that it's guaranteed that built-in sidecar containers finish AFTER all the regular containers completion may not be always true.

SergeyKanzhelev replied:
rephrased

sftim commented:
My key feedback is (still) the “questions in my mind” from https://github.com/kubernetes/website/pull/46825#pullrequestreview-2134319796

afro-coder commented on content/en/docs/tutorials/configuration/pod-sidecar-containers.md:
```diff
@@ -0,0 +1,148 @@
+   built-in sidecar containers will not block Pod completion.
+4. Also, with Jobs, build-in sidecar containers will keep being restarted, even if regular containers would not with Pod's `restartPolicy=Never`.
+
+These benefits are drastic, and lead to faster adoption that we see with other Kubernetes features.
```
Trying to help with this article, I tried going through the 3 pages we have

[Locally]
https://kubernetes.io/docs/tutorials/configuration/pod-sidecar-containers/

[Active]
https://kubernetes.io/blog/2023/08/25/native-sidecar-containers/
https://kubernetes.io/docs/concepts/workloads/pods/sidecar-containers/

Definitely a bit confusing, I'm also trying to contribute, As of now there is `tutorial` in this article, to begin with I'll be adding an example of how one can enable sidecars for something like maybe echoserver.

Additionally I'm a bit confused, do I just clone this repo with the `sidecarsTrobleshooting` branch and then push a commit to that? or do I fork the repo and create a branch and then merge it with SergeyKanzhelev:sidecarsTrobleshooting?

SergeyKanzhelev replied:
removed the statement

T-Lakshmi commented:
/label tide/merge-method-squash

@SergeyKanzhelev,
can you please squash the commits.

SergeyKanzhelev commented on content/en/docs/tutorials/configuration/pod-sidecar-containers.md:
```diff
@@ -0,0 +1,148 @@
+
+These benefits are drastic, and lead to faster adoption that we see with other Kubernetes features.
+
+## Adopting sidecar containers
```
I changed the title. Perhaps the old way can be added as a next PR

SergeyKanzhelev commented on content/en/docs/tutorials/configuration/pod-sidecar-containers.md:
```diff
@@ -0,0 +1,148 @@
+
+Here are the considerations and troubleshooting steps that one can take while adopting sidecar containers for their workload.
+
+**Ensure the feature gate is enabled.**
```
only two items in this list. Fixed

SergeyKanzhelev commented on content/en/docs/tutorials/configuration/pod-sidecar-containers.md:
```diff
@@ -0,0 +1,148 @@
+**Ensure the feature gate is enabled.**
+
+As a very first step, make sure that both - API server and Nodes were upgraded to the version `1.29+`.
+The feature will break on clusters where Nodes are running earlier versions where it is not enabled.
```
added an alert

SergeyKanzhelev commented on content/en/docs/tutorials/configuration/pod-sidecar-containers.md:
```diff
@@ -0,0 +1,148 @@
+All of the strategies are generally options you may choose to decide whether
+the Pod the sidecar will be injected to will land on a Node supporting the feature or not.
+
+As an example, you can follow this conversation: https://github.com/istio/istio/issues/48794
```
fixed

SergeyKanzhelev commented on content/en/docs/tutorials/configuration/pod-sidecar-containers.md:
```diff
@@ -0,0 +1,148 @@
+     detect the node version and exit immediately if the version does not support the sidecar feature.
+   - Consider runtime feature detection design:
+     - Define an empty dir so containers can communicate with each other
+     - Inject init container `NativeSidecar` with `restartPolicy=Always`. 
```
fixed

sftim commented:
/remove-label tide/merge-method-squash

sftim commented on content/en/docs/tutorials/configuration/pod-sidecar-containers.md:
```diff
@@ -0,0 +1,166 @@
+<!-- overview -->
+
+This section is relevant for people adopting [sidecar containers](/docs/concepts/workloads/pods/sidecar-containers/) for their workloads.
+
```
We should (I think need to) explain that you can use sidecars with older clusters. Otherwise, we risk seriously misleading readers.

SergeyKanzhelev replied:
added a paragraph

sftim commented on content/en/docs/tutorials/configuration/pod-sidecar-containers.md:
```diff
@@ -0,0 +1,166 @@
+
+## Benefits of a built-in sidecar containers
+
+There are many benefits for the built-in sidecar containers.
```
```suggestion
Using Kubernetes' native support for sidecar containers provides several benefits:
```

sftim commented on content/en/docs/tutorials/configuration/pod-sidecar-containers.md:
```diff
@@ -0,0 +1,166 @@
+
+There are many benefits for the built-in sidecar containers.
+
+1. Built-in sidecar containers can run before Init containers.
```
```suggestion
1. You can configure a native  sidecar container to start ahead of {{< glossary_tooltip text="init containers" term_id="init-container" >}}.
```

sftim commented on content/en/docs/tutorials/configuration/pod-sidecar-containers.md:
```diff
@@ -0,0 +1,166 @@
+   are completed and terminated. If the sidecar container isn’t gracefully shut down, a 
+   `SIGKILL` signal will be used to terminate it.
+3. With Jobs, when Pod's `restartPolicy: OnFailure` or `restartPolicy: Never`,
+   built-in sidecar containers will not block Pod completion.
```
```suggestion
   native sidecar containers do not block Pod completion. With legacy sidecar containers,
   special care is needed to handle this situation.
```

sftim reviewed (commented):
Some more feedback. Please address the existing feedback as well.

SergeyKanzhelev commented on content/en/docs/tutorials/configuration/pod-sidecar-containers.md:
```diff
@@ -0,0 +1,148 @@
+
+{{< /alert >}}
+
+**Automatic injection of sidecars**
```
changed to headers

SergeyKanzhelev commented:
@sftim I think I addressed most of comments. I really want this first step out. See https://kubernetes.slack.com/archives/C0EG7JC6T/p1723141448471749

I added a few follow up steps on top that can be done. @afro-coder I think it will be easiest if we can merge this PR and if you want to contribute, can do as a follow up.

matthyx commented:
great write, 100% correct regarding technicalities, thanks Sergey!
/lgtm

tengqm commented:
/approve
Let's kick this in and iterate over it.
